File: Blockchain.py

#instantiate a block
class BlockObj:
    def __init__(self, data, id, prev):
        self.data = data
        self.id = id
        self.prev = prev

    #Return the data of the block
    def blockObj_data(self):
        return self.data

    #return previos hash
    def blockObj_prev(self):
        return self.prev

    #compute the hash
    def hash(self):
        add = self.id + self.prev
        for i in range(len(self.data)):
            add += int(self.data[i])
        return add

    #print out the contents of the block
    def printBlock(self):
        print(self.id)
        print(self.data)

class Blockchain():
    def __init__(self):
        self.blocks = []

    #Return list of blocks
    def getBlock(self, i):
        return self.blocks[i]

    #Return number of elements in blockchain
    def size(self):
        return len(self.blocks)

    #Add new block to chain
    def appendBlock(self, text):
        if len(self.blocks) == 0:
            self.blocks.append(BlockObj(text, len(self.blocks), 0))
        else:
            self.blocks.append(BlockObj(text, len(self.blocks), hash( self.blocks[len(self.blocks)-1])))

    def removeLast(self):
        del(self.blocks[-1])

    #print out the contents of the blockchain
    def printBlockchain(self):
        for i in self.blocks:
            i.printBlock()


obj = Blockchain()
obj.appendBlock("whatever")

obj.appendBlock("threee")
obj.appendBlock("nope")

obj.printBlockchain()

File: Blockchain.go
package main

import ("fmt")
// import "runtime/debug"

type Block struct { 
	data string 
	id  int
	prev float64  
}

// type BlockFuctions interface {
//     getData() string
// 	previosHash() float64
// 	hash() float64
// 	printBlock() 
// 	newBlock() Block
	
// }

func (b Block) getData() string {
    return b.data
}

func (b Block) previosHash() float64 {
    return b.prev
}


func (b Block) hash() float64 {
	var total = float64(b.id) + b.prev

   for i := 0; i < len(b.data); i++ {
	   total=total+float64(b.data[i])
	} 

	return total
}

func (b Block) printBlock() {

	fmt.Print(b.id)
	fmt.Print(" ")
	fmt.Print(b.data)
	fmt.Print(" ")
}

type Blockchain struct { 
	count int
    chain []Block
}

// type BlockchainFuctions interface {
// 	get() Block
// 	size() int
// 	appendBlock()
// 	removeLast()
// 	printBlockchain()

	
// }

func (B Blockchain) get(id int) Block {
	return (B.chain[id])
}

func (B Blockchain) size() int {
	return (B.count)
}

func (B *Blockchain) appendBlock(data string) {


	if(B.count==0){

		var tempBlock = Block{data, B.count, 0}
		B.chain=append(B.chain, tempBlock)

	}else{

		var tempBlock = Block{data, B.count, B.chain[(B.count)-1].hash()}
		B.chain=append(B.chain, tempBlock)
	}

	B.count=B.count+1
	
}



func (B *Blockchain) removeLast() {
	if B.count==0 {
		return
	}else {
		B.chain = B.chain[:len(B.chain)-1] 
		B.count=B.count-1
	}
		
}


func (B Blockchain) printBlockchain() {
	
	for i := 0; i < B.count; i++ {
		B.chain[i].printBlock()	
	}
	fmt.Print("\n")
}


func main(){
	var myChain=Blockchain{count:0}

	// var block1 = Block{"hi",2,3}
	// fmt.Print(block1.getData())
	
	myChain.appendBlock("one")

	

	myChain.appendBlock("two")



	// fmt.Print(myChain.chain)

	// fmt.Print(myChain.count)


	
	myChain.printBlockchain()
}


